ca/certificate-authority.go

```diff
@@ -216,6 +272,57 @@ func (ca *CertificateAuthorityImpl) noteHSMFault(err error) {
 	return
 }
 
+// Extract supported extensions from a CSR.  The following extensions are
+// currently supported:
+//
+// * 1.3.6.1.5.5.7.1.24 - TLS Feature [RFC7633], with the "must staple" value.
+//                        Any other value will result in an error.
+//
+// Other requested extensions are silently ignored.
+func (ca *CertificateAuthorityImpl) extensionsFromCSR(csr *x509.CertificateRequest) (extensions []signer.Extension, err error) {
+	extensions = []signer.Extension{}
+	for _, attr := range csr.Attributes {
+		if !attr.Type.Equal(oidExtensionRequest) {
+			continue
+		}
+
+		for _, extList := range attr.Value {
+			for _, ext := range extList {
+				switch {
+				case ext.Type.Equal(oidTLSFeature):
+					ca.stats.Inc(metricCSRExtensionTLSFeature, 1, 1.0)
+					value, ok := ext.Value.([]byte)
+					if !ok {
+						msg := fmt.Sprintf("Mal-formed extension with OID %v", ext.Type)
+						err = core.CertificateIssuanceError(msg)
+						return
+					} else if !bytes.Equal(value, mustStapleFeatureValue) {
+						msg := fmt.Sprintf("Unsupported value for extension with OID %v", ext.Type)
+						ca.stats.Inc(metricCSRExtensionTLSFeatureInvalid, 1, 1.0)
+						err = core.CertificateIssuanceError(msg)
+						return
+					}
+
+					extensions = append(extensions, mustStapleExtension)
+				case ext.Type.Equal(oidAuthorityInfoAccess),
+					ext.Type.Equal(oidAuthorityKeyIdentifier),
+					ext.Type.Equal(oidBasicConstraints),
+					ext.Type.Equal(oidCertificatePolicies),
+					ext.Type.Equal(oidCrlDistributionPoints),
+					ext.Type.Equal(oidExtKeyUsage),
+					ext.Type.Equal(oidKeyUsage),
+					ext.Type.Equal(oidSubjectAltName),
+					ext.Type.Equal(oidSubjectKeyIdentifier):
+					ca.stats.Inc(metricCSRExtensionBasic, 1, 1.0)
```

Comment:
This will increment once for each basic extension in the CSR, which doesn't quite match the comment "Increments when CA handles a CSR requesting a "basic" extension."

I'm not sure what value we get from this metric, either. I would probably just drop it.